### PR TITLE
App extensions

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Renderer and editor

### Description
Besides some refactoring this PR introduces the basic app extension class which is used to easily add new markdown rendering integrations, code mirror linters, etc. It is not planned to provide a public, fully documented api. It's only used internally to organize the code.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [ ] Added / updated documentation
- [ ] Extended changelog
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### CODE REVIEW

File: src/components/editor-page/renderer-pane/render-iframe.tsx [hunk 1]
```diff
@@ -105,15 +107,15 @@ export const RenderIframe: React.FC<RenderIframeProps> = ({
...
-        const lineOffset = getGlobalState().noteDetails.frontmatterRendererInfo.lineOffset
-        onTaskCheckedChange?.(values.lineInMarkdown + lineOffset, values.checked)
-      },
-      [onTaskCheckedChange]
-    )
+    CommunicationMessageType.EXTENSION_EVENT,
+    useMemo(() => {
+      return eventEmitter === undefined
+        ? undefined
+        : (values: ExtensionEvent) => eventEmitter?.emit(values.eventName, values.payload)
```
Comment (DerMolly) [inline src/components/editor-page/renderer-pane/render-iframe.tsx:117]:
```suggestion
        : (values: ExtensionEvent) => eventEmitter!.emit(values.eventName, values.payload)
```
At this point we know eventEmitter is not undefined

File: src/components/editor-page/editor-pane/hooks/use-components-from-app-extensions.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import type { ReactElement } from 'react'
+import React, { Fragment, useMemo } from 'react'
+import { extraAppExtensions } from '../../../../extensions/extra-integrations/extra-app-extensions'
+
+export const useComponentsFromAppExtensions = (): ReactElement => {
```
Comment (ErikMichelson) [inline src/components/editor-page/editor-pane/hooks/use-components-from-app-extensions.tsx:11]: ESDoc

File: src/components/markdown-renderer/hooks/use-markdown-extensions.ts [hunk 3]
```diff
@@ -5,98 +5,57 @@
...
-
+  additionalExtensions: MarkdownRendererExtension[]
+): MarkdownRendererExtension[] => {
+  const extensionEventEmitter = useExtensionEventEmitter()
+  //replace with global list
   return useMemo(() => {
     return [
-      new TableOfContentsMarkdownExtension((ast?: TocAst) => (toc.current = ast)),
+      ...extraRendererExtensions,
       ...additionalExtensions,
```
Comment (DerMolly) [inline src/components/markdown-renderer/hooks/use-markdown-extensions.ts:45]: I find the difference between `extraRendererExtensions` and `additionalExtensions` a bit puzzeling…

File: src/components/markdown-renderer/extensions/image/event-emitting-proxy-image-frame.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import React, { useCallback } from 'react'
+import { useExtensionEventEmitter } from '../../hooks/use-extension-event-emitter'
+import { ProxyImageFrame } from './proxy-image-frame'
+import type { ImageDetails } from '../../../render-page/window-post-message-communicator/rendering-message'
+
+type EventEmittingProxyImageFrameProps = Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'onClick'>
+
+export const EVENT_NAME = 'ImageClick'
+
+export const EventEmittingProxyImageFrame: React.FC<EventEmittingProxyImageFrameProps> = (props) => {
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/extensions/image/event-emitting-proxy-image-frame.tsx:16]: ESDoc

File: src/components/markdown-renderer/extensions/image/event-emitting-proxy-image-frame.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+import React, { useCallback } from 'react'
+import { useExtensionEventEmitter } from '../../hooks/use-extension-event-emitter'
+import { ProxyImageFrame } from './proxy-image-frame'
+import type { ImageDetails } from '../../../render-page/window-post-message-communicator/rendering-message'
+
+type EventEmittingProxyImageFrameProps = Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'onClick'>
+
+export const EVENT_NAME = 'ImageClick'
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/extensions/image/event-emitting-proxy-image-frame.tsx]: Move to class

File: src/components/render-page/markdown-document.tsx [hunk 6]
```diff
@@ -106,18 +97,14 @@ export const MarkdownDocument: React.FC<MarkdownDocumentProps> = ({
...
-          onImageClick={onImageClick}
           newlinesAreBreaks={newlinesAreBreaks}
         />
       </div>
-      <div className={`${styles['markdown-document-side']} pt-4`}>
-        <ShowIf condition={!!tocAst && !disableToc}>
-          <WidthBasedTableOfContents tocAst={tocAst as TocAst} baseUrl={baseUrl} width={containerWidth} />
-        </ShowIf>
-      </div>
+      <DocumentTocSidebar
```
Comment (ErikMichelson) [inline src/components/render-page/markdown-document.tsx:104]: Self-closable tag

File: src/extensions/extra-integrations/alert/alert-app-extension.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import { AlertMarkdownExtension } from './alert-markdown-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+
+export class AlertAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/alert/alert-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/extra-app-extensions.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import { LegacySequenceDiagramAppExtension } from './sequence-diagram/legacy-sequence-diagram-app-extension'
+import { SpoilerAppExtension } from './spoiler/spoiler-app-extension'
+import { VegaLiteAppExtension } from './vega-lite/vega-lite-app-extension'
+import { VimeoAppExtension } from './vimeo/vimeo-app-extension'
+import { YoutubeAppExtension } from './youtube/youtube-app-extension'
+import { HighlightedCodeFenceAppExtension } from './highlighted-code-fence/highlighted-code-fence-app-extension'
+import type { AppExtension } from '../base/app-extension'
+import { TaskListCheckboxAppExtension } from './task-list/task-list-checkbox-app-extension'
+
+export const extraAppExtensions: AppExtension[] = [
```
Comment (DerMolly) [inline src/extensions/extra-integrations/extra-app-extensions.ts]: This should probably be named `optionalAppExtensions`

File: src/extensions/extra-integrations/blockquote/blockquote-app-extension.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { BlockquoteExtraTagMarkdownExtension } from './blockquote-extra-tag-markdown-extension'
+
+export class BlockquoteAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/blockquote/blockquote-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/csv/csv-table-app-extension.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { CsvTableMarkdownExtension } from './csv-table-markdown-extension'
+
+export class CsvTableAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/csv/csv-table-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/extra-app-extensions.ts (hunk 8 above)
Comment (ErikMichelson) [inline src/extensions/extra-integrations/extra-app-extensions.ts]: ESDoc + naming

File: src/extensions/extra-integrations/flowchart/flowchart-app-extension.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { FlowchartMarkdownExtension } from './flowchart-markdown-extension'
+
+export class FlowchartAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/flowchart/flowchart-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/gist/gist-app-extension.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { GistMarkdownExtension } from './gist-markdown-extension'
+
+export class GistAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/gist/gist-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/graphviz/graphviz-app-extension.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { GraphvizMarkdownExtension } from './graphviz-markdown-extension'
+
+export class GraphvizAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/graphviz/graphviz-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/highlighted-code-fence/highlighted-code-fence-app-extension.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { HighlightedCodeMarkdownExtension } from './highlighted-code-markdown-extension'
+
+export class HighlightedCodeFenceAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/highlighted-code-fence/highlighted-code-fence-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/katex/katex-app-extension.ts [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { KatexMarkdownExtension } from './katex-markdown-extension'
+
+export class KatexAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/katex/katex-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/legacy-short-codes/legacy-shortcodes-app-extension.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { LegacyShortcodesMarkdownExtension } from './legacy-shortcodes-markdown-extension'
+import type { Linter } from '../../../components/editor-page/editor-pane/linter/linter'
+import { SingleLineRegexLinter } from '../../../components/editor-page/editor-pane/linter/single-line-regex-linter'
+import { legacySpeakerdeckRegex } from './replace-legacy-speakerdeck-short-code'
+import { t } from 'i18next'
+import { legacySlideshareRegex } from './replace-legacy-slideshare-short-code'
+import { legacyPdfRegex } from './replace-legacy-pdf-short-code'
+
+export class LegacyShortcodesAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/legacy-short-codes/legacy-shortcodes-app-extension.ts:16]: ESDoc

**Review by DerMolly — CHANGES_REQUESTED** [review]

File: src/extensions/extra-integrations/mermaid/mermaid-app-extension.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { MermaidMarkdownExtension } from './mermaid-markdown-extension'
+
+export class MermaidAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/mermaid/mermaid-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/plantuml/plantuml-app-extension.ts [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { PlantumlMarkdownExtension } from './plantuml-markdown-extension'
+
+export class PlantumlAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/plantuml/plantuml-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/sequence-diagram/legacy-sequence-diagram-app-extension.ts [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { LegacySequenceDiagramMarkdownExtension } from './legacy-sequence-diagram-markdown-extension'
+import type { Linter } from '../../../components/editor-page/editor-pane/linter/linter'
+import { SingleLineRegexLinter } from '../../../components/editor-page/editor-pane/linter/single-line-regex-linter'
+import { t } from 'i18next'
+
+export class LegacySequenceDiagramAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/sequence-diagram/legacy-sequence-diagram-app-extension.ts:14]: ESDoc

File: src/extensions/extra-integrations/spoiler/spoiler-app-extension.ts [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { SpoilerMarkdownExtension } from './spoiler-markdown-extension'
+
+export class SpoilerAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/spoiler/spoiler-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/task-list/event-emitting-task-list-checkbox.tsx [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+type EventEmittingTaskListCheckboxProps = Omit<TaskListProps, 'onTaskCheckedChange' | 'disabled'>
+
+export const EVENT_NAME = 'TaskListCheckbox'
+
+export interface TaskCheckedEventPayload {
+  lineInMarkdown: number
+  checked: boolean
+}
+
+export const EventEmittingTaskListCheckbox: React.FC<EventEmittingTaskListCheckboxProps> = (props) => {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/task-list/event-emitting-task-list-checkbox.tsx:20]: ESDoc

File: src/components/markdown-renderer/hooks/use-markdown-extensions.ts (hunk 3 above)
Comment (mrdrogdrog) [reply src/components/markdown-renderer/hooks/use-markdown-extensions.ts:45]: I renamed the extra extensions to optional.

File: src/extensions/extra-integrations/task-list/set-checkbox-in-editor.tsx [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import type React from 'react'
+import { useSetCheckboxInEditor } from './use-set-checkbox-in-editor'
+import { useExtensionEventEmitterHandler } from '../../../components/markdown-renderer/hooks/use-extension-event-emitter'
+import { EVENT_NAME } from './event-emitting-task-list-checkbox'
+
+export const SetCheckboxInEditor: React.FC = () => {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/task-list/set-checkbox-in-editor.tsx:12]: ESDoc

File: src/extensions/extra-integrations/task-list/task-list-checkbox-app-extension.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import type EventEmitter2 from 'eventemitter2'
+import { AppExtension } from '../../base/app-extension'
+import { TaskListMarkdownExtension } from './task-list-markdown-extension'
+import type React from 'react'
+import { SetCheckboxInEditor } from './set-checkbox-in-editor'
+
+export class TaskListCheckboxAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/task-list/task-list-checkbox-app-extension.ts:13]: ESDoc

File: src/extensions/extra-integrations/vega-lite/vega-lite-app-extension.ts [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { VegaLiteMarkdownExtension } from './vega-lite-markdown-extension'
+
+export class VegaLiteAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/vega-lite/vega-lite-app-extension.ts:11]: ESDoc

File: src/extensions/extra-integrations/vimeo/vimeo-app-extension.ts [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { VimeoMarkdownExtension } from './vimeo-markdown-extension'
+import type { Linter } from '../../../components/editor-page/editor-pane/linter/linter'
+import { SingleLineRegexLinter } from '../../../components/editor-page/editor-pane/linter/single-line-regex-linter'
+import { legacyVimeoRegex } from './replace-legacy-vimeo-short-code'
+import { t } from 'i18next'
+
+export class VimeoAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/vimeo/vimeo-app-extension.ts:15]: ESDoc

File: src/extensions/extra-integrations/youtube/youtube-app-extension.ts [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+
+import { AppExtension } from '../../base/app-extension'
+import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
+import { YoutubeMarkdownExtension } from './youtube-markdown-extension'
+import type { Linter } from '../../../components/editor-page/editor-pane/linter/linter'
+import { SingleLineRegexLinter } from '../../../components/editor-page/editor-pane/linter/single-line-regex-linter'
+import { legacyYouTubeRegex } from './replace-legacy-youtube-short-code'
+import { t } from 'i18next'
+
+export class YoutubeAppExtension extends AppExtension {
```
Comment (ErikMichelson) [inline src/extensions/extra-integrations/youtube/youtube-app-extension.ts:15]: ESDoc

**Review by ErikMichelson — CHANGES_REQUESTED** [review]

**Review by ErikMichelson — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Pull request merged**